src/QsCompiler/CompilerOptimizations/ConstantPropagation.fs

```diff
@@ -27,101 +29,129 @@ type ConstantPropagator(compiledCallables: ImmutableDictionary<QsQualifiedName,
-                            {   Statement = innerScope |> QsScopeStatement.New |> QsScopeStatement
-                                SymbolDeclarations = stm.Body.KnownSymbols
-                                Location = Null
-                                Comments = QsComments.New ([], []) })
-                    let outerScope = QsScope.New (iterRange, stm.Body.KnownSymbols)
-                    let result = outerScope |> QsScopeStatement.New |> QsScopeStatement |> statementKind.Transform
-                    // printfn "Unrolled for loop!\nOld statement was:\n%O\nNew statement is:\n%O" (printStm 0 (QsForStatement stm)) (printStm 0 result)
-                    result
-                | None ->
-                    let loopVarType = statementKind.TypeTransformation (snd stm.LoopItem)
-                    let body = statementKind.ScopeTransformation stm.Body
-                    QsForStatement.New ((loopVar, loopVarType), iterVals, body) |> QsForStatement
         }
     }
+
+
+/// The StatementKindTransformation used to simplify statements
+and [<AbstractClass>] private StatementOptimizer(stateRef: TransformationState ref) =
+    inherit StatementKindTransformation()
+
+
+    override this.onExpressionStatement ex =
+        let ex = this.ExpressionTransformation ex
+        maybe {
+            let! qualName, newScope = tryInline !stateRef ex
+            let mySet = HashSet()
+            findAllCalls !stateRef newScope mySet
+            let! currentCallable = (!stateRef).currentCallable
+            do! not (mySet.Contains currentCallable.FullName) |> check
+            let s1 = printNamespaceElem (QsCallable currentCallable)
+            let s2 = printScope 0 newScope
+            let newScope2 = this.ScopeTransformation newScope
+            return newScope2 |> QsScopeStatement.New |> QsScopeStatement
+            // return newScope |> QsScopeStatement.New |> QsScopeStatement
+        } |? QsExpressionStatement ex
+
+
+    override this.onConditionalStatement stm = 
+        let cbList, cbListEnd =
+            stm.ConditionalBlocks |> Seq.fold (fun s (cond, block) ->
+                let newCond = this.ExpressionTransformation cond
+                match newCond.Expression with
+                | BoolLiteral true -> s @ [None, block]
+                | BoolLiteral false -> s
+                | _ -> s @ [Some cond, block]
+            ) [] |> List.ofSeq |> takeWhilePlus1 (fun (c, _) -> c <> None)
+        let newDefault = cbListEnd |> Option.map (snd >> Value) |? stm.Default
+
+        let cbList = cbList |> List.map (fun (c, b) -> this.onPositionedBlock (c, b))
+        let newDefault = match newDefault with Value x -> this.onPositionedBlock (None, x) |> snd |> Value | Null -> Null
+
+        match cbList, newDefault with
+        | [], Value x ->
+            x.Body |> QsScopeStatement.New |> QsScopeStatement
+        | [], Null ->
+            QsScope.New ([], LocalDeclarations.New []) |> QsScopeStatement.New |> QsScopeStatement
```

Comment:
What is your progress in getting rid of the Scope statement you introduced?

Reply:
Fixed in future PR